File: week-1/friend-circles/friend-circles.js
/**
 * @param {number[][]} M
 * @return {number}
 */
var findCircleNum = function(M) {
    let count = 0;
    let visited = {};
    for (let i=0; i<M.length; i++) {
        if (!visited.hasOwnProperty(i)) {
            dfs(i, visited, M);
            ++count;
        }
    }
    
    return count;
};

const dfs = function (row, visited, M) {
    for (let i=0; i<M[row].length; i++) {
        if (M[row][i] === 1 && !visited[i]) {
            visited[i] = true;
            dfs(i, visited, M);
        }
    }
}

File: week-1/friend-circles/friend-circles.py
class Solution:
    def findCircleNum(self, M: List[List[int]]) -> int:
        visited = {}
        count = 0
        for i in range(len(M)):
            if i not in visited:
                self.dfs(i, visited, M)
                count += 1
                
        return count
    
    def dfs(self, row, visited, M):
        for key, val in enumerate(M[row]):
            if val == 1 and key not in visited:
                visited[key] = True
                self.dfs(key, visited, M)
